{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import argparse\n",
    "# import os, sys\n",
    "# sys.path.append(os.path.abspath('..'))\n",
    "# import test_metric\n",
    "\n",
    "import json\n",
    "import requests\n",
    "\n",
    "import rdflib\n",
    "from rdflib import ConjunctiveGraph\n",
    "import pyshacl\n",
    "\n",
    "import extruct\n",
    "from extruct.jsonld import JsonLdExtractor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of URIs to be tested\n",
    "# http://lod.openaire.eu/documentation \n",
    "\n",
    "input_URIs = ['https://data.inra.fr/dataset.xhtml?persistentId=doi:10.15454/TKMGCQ', \n",
    "    'https://doi.org/10.5281/zenodo.1147435',\n",
    "    'https://data.inra.fr/dataset.xhtml?persistentId=doi:10.15454/5K9HCS', \n",
    "    'https://search.datacite.org/works/10.7892/boris.108387', \n",
    "    'https://doi.org/10.6084/m9.figshare.c.3607916_d7.v1', \n",
    "    'https://search.datacite.org/works/10.6084/m9.figshare.c.3607916_d7.v1', \n",
    "    'https://api.datacite.org/dois/application/ld+json/10.6084/m9.figshare.c.3607916_d7.v1', \n",
    "    'https://bio.tools/rsat_peak-motifs',\n",
    "    'https://workflowhub.eu/workflows/18', \n",
    "    'https://tess.elixir-europe.org/materials/train-the-trainer', \n",
    "    'https://fairsharing.org/FAIRsharing.ZPRtfG', \n",
    "    'https://www.ebi.ac.uk/ols/ontologies/ncit/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FNCIT_C2985', \n",
    "    'https://biit.cs.ut.ee/gprofiler/page/apis', \n",
    "    'https://bioportal.bioontology.org/ontologies/OCRE', \n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input of URIs by type\n",
    "\n",
    "input_data = [\n",
    "    \"https://data.inra.fr/dataset.xhtml?persistentId=doi:10.15454/TKMGCQ\", # dataset INRA Dataverse\n",
    "    \"https://doi.pangaea.de/10.1594/PANGAEA.914331\", # dataset in PANGAEA\n",
    "]\n",
    "\n",
    "input_software = [\n",
    "    \"https://zenodo.org/record/3349821#.Xp7m9SNR2Uk\", # VM image in zenodo\n",
    "    \"https://explore.openaire.eu/search/software?softwareId=r37b0ad08687::275ecd99e516ed1b863e2a7586063a64\", # same VM image in OpenAir\n",
    "    \"https://data.inra.fr/dataset.xhtml?persistentId=doi:10.15454/5K9HCS\", # code in INRA Dataverse\n",
    "    \"https://bio.tools/rsat_peak-motifs\", # Tool in biotools\n",
    "    \"https://workflowhub.eu/workflows/18\", # Workflow in WorkflowHub\n",
    "    \"http://tara-oceans.mio.osupytheas.fr/ocean-gene-atlas/\", # OGA Main page of webtool\n",
    "]\n",
    "\n",
    "input_database = [\n",
    "    \"https://fairsharing.org/FAIRsharing.ZPRtfG\", # knowledge base in FAIRsharing (AgroLD)\n",
    "    \"http://remap.univ-amu.fr\" # Database of transcriptional regulators\n",
    "]\n",
    "\n",
    "input_ontology = [\n",
    "    \"https://bioportal.bioontology.org/ontologies/OCRE\" # Ontology in bioportal\n",
    "    \"https://www.ebi.ac.uk/ols/ontologies/ncit/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FNCIT_C2985\" # OLS entry\n",
    "]\n",
    "\n",
    "input_publication = [\n",
    "    \"https://search.datacite.org/works/10.7892/boris.108387\", # Publication in Datacite\n",
    "    \"https://doi.org/10.6084/m9.figshare.c.3607916_d7.v1\", # Publication figure in FigShare\n",
    "    \"https://search.datacite.org/works/10.6084/m9.figshare.c.3607916_d7.v1\", # Publication figure in Datacite (same as previous)\n",
    "    \"https://api.datacite.org/dois/application/ld+json/10.6084/m9.figshare.c.3607916_d7.v1\" # Publication figure with Datacite API\n",
    "]\n",
    "\n",
    "input_training = [\n",
    "    \"https://tess.elixir-europe.org/materials/train-the-trainer\", # Training material in TeSS\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Retrieving of RDF linked-data in web pages\n",
    "\n",
    "# for URI in input_URIs:\n",
    "# for URI in input_training:\n",
    "# for URI in input_publication:\n",
    "# for URI in input_ontology:\n",
    "# for URI in input_database:\n",
    "def getKG():\n",
    "    for URI in input_software:\n",
    "        print(URI)\n",
    "        page = requests.get(URI)\n",
    "        html = page.content\n",
    "        d = extruct.extract(html, syntaxes=['microdata', 'rdfa', 'json-ld'], errors='ignore')\n",
    "        #print(json.dumps(data, indent=2))\n",
    "        kg = ConjunctiveGraph()\n",
    "        kg.parse(data = json.dumps(d['json-ld']), format=\"json-ld\")\n",
    "        kg.parse(data = json.dumps(d['rdfa']), format=\"json-ld\")\n",
    "        kg.parse(data = json.dumps(d['microdata']), format=\"json-ld\")\n",
    "        #print(kg.serialize(format=\"turtle\").decode())\n",
    "        #break\n",
    "        return kg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Augmenting embedded semantic annotation from multiple sources\n",
    "1. for each ressource find all typed or identified entities \n",
    "2. for each entity: retrieve their semantic description based on other SPARQL endpoints. 'DESCRIBE queries' \n",
    "   - datacite sparql endpoint\n",
    "   - sameAs.cc sparql endpoint\n",
    "   - BioPortal sparql endpoint\n",
    "   - OpenAire sparql endpoint\n",
    "   - PubMed sparql endpoint (http://linkedlifedata.com/sparql)\n",
    "   - DBPedia\n",
    "   - WikiData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://zenodo.org/record/3349821#.Xp7m9SNR2Uk\n",
      "https://doi.org/10.5281/zenodo.3349821 - http://schema.org/identifier - https://doi.org/10.5281/zenodo.3349821\n"
     ]
    }
   ],
   "source": [
    "from SPARQLWrapper import SPARQLWrapper, N3\n",
    "from rdflib import Graph, Namespace\n",
    "from rdflib.namespace import RDF\n",
    "\n",
    "schema = Namespace(\"http://schema.org/\")\n",
    "\n",
    "#sparql_endpoints = ['http://linkedlifedata.com/sparql', 'http://lod.openaire.eu/sparql']\n",
    "sparql_endpoints = ['http://lod.openaire.eu/sparql', 'http://sparql.bioontology.org/ontologies/sparql']\n",
    "\n",
    "def retrieve_entities(G):\n",
    "    entities = []\n",
    "    for s, p, o in G.triples((None, RDF.type, None)):\n",
    "        entities.append(str(s))\n",
    "        #print(f\"{s}\")\n",
    "    for s, p, o in G.triples((None, schema.identifier, None)):\n",
    "        print(f'{s} - {p} - {o}')\n",
    "        entities.append(str(o))\n",
    "    return entities\n",
    "\n",
    "entities = retrieve_entities(getKG())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "searching for http://edamontology.org/topic_3307\n"
     ]
    },
    {
     "ename": "HTTPError",
     "evalue": "HTTP Error 403: FORBIDDEN",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mHTTPError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-48-1fcb59fbf96f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0muri\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mentities\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'searching for {uri}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 18\u001b[0;31m     \u001b[0maugmented_kg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdescribe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0muri\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     19\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maugmented_kg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mserialize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'turtle'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdecode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-48-1fcb59fbf96f>\u001b[0m in \u001b[0;36mdescribe\u001b[0;34m(uri)\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m         \u001b[0msparql\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msetReturnFormat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mN3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m         \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msparql\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconvert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m         \u001b[0mg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"n3\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/biotools-rdf/lib/python3.6/site-packages/SPARQLWrapper/Wrapper.py\u001b[0m in \u001b[0;36mquery\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    796\u001b[0m             \u001b[0;34m@\u001b[0m\u001b[0mrtype\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mL\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0mQueryResult\u001b[0m\u001b[0;34m}\u001b[0m \u001b[0minstance\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    797\u001b[0m         \"\"\"\n\u001b[0;32m--> 798\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mQueryResult\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_query\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    799\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    800\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mqueryAndConvert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/biotools-rdf/lib/python3.6/site-packages/SPARQLWrapper/Wrapper.py\u001b[0m in \u001b[0;36m_query\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    776\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mEndPointInternalError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    777\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 778\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    779\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    780\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mquery\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/biotools-rdf/lib/python3.6/site-packages/SPARQLWrapper/Wrapper.py\u001b[0m in \u001b[0;36m_query\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    764\u001b[0m                 \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0murlopener\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    765\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 766\u001b[0;31m                 \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0murlopener\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    767\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturnFormat\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    768\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0murllib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merror\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mHTTPError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/biotools-rdf/lib/python3.6/urllib/request.py\u001b[0m in \u001b[0;36murlopen\u001b[0;34m(url, data, timeout, cafile, capath, cadefault, context)\u001b[0m\n\u001b[1;32m    221\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    222\u001b[0m         \u001b[0mopener\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_opener\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 223\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mopener\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    224\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    225\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0minstall_opener\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mopener\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/biotools-rdf/lib/python3.6/urllib/request.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[1;32m    530\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mprocessor\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprocess_response\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprotocol\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    531\u001b[0m             \u001b[0mmeth\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocessor\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 532\u001b[0;31m             \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmeth\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    533\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    534\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/biotools-rdf/lib/python3.6/urllib/request.py\u001b[0m in \u001b[0;36mhttp_response\u001b[0;34m(self, request, response)\u001b[0m\n\u001b[1;32m    640\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m200\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mcode\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m300\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    641\u001b[0m             response = self.parent.error(\n\u001b[0;32m--> 642\u001b[0;31m                 'http', request, response, code, msg, hdrs)\n\u001b[0m\u001b[1;32m    643\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    644\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/biotools-rdf/lib/python3.6/urllib/request.py\u001b[0m in \u001b[0;36merror\u001b[0;34m(self, proto, *args)\u001b[0m\n\u001b[1;32m    568\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mhttp_err\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    569\u001b[0m             \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mdict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'default'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'http_error_default'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0morig_args\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 570\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_chain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    571\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    572\u001b[0m \u001b[0;31m# XXX probably also want an abstract factory that knows when it makes\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/biotools-rdf/lib/python3.6/urllib/request.py\u001b[0m in \u001b[0;36m_call_chain\u001b[0;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[1;32m    502\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mhandler\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mhandlers\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    503\u001b[0m             \u001b[0mfunc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandler\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 504\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    505\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    506\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/biotools-rdf/lib/python3.6/urllib/request.py\u001b[0m in \u001b[0;36mhttp_error_default\u001b[0;34m(self, req, fp, code, msg, hdrs)\u001b[0m\n\u001b[1;32m    648\u001b[0m \u001b[0;32mclass\u001b[0m \u001b[0mHTTPDefaultErrorHandler\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mBaseHandler\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    649\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mhttp_error_default\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmsg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhdrs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 650\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mHTTPError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfull_url\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmsg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhdrs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    651\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    652\u001b[0m \u001b[0;32mclass\u001b[0m \u001b[0mHTTPRedirectHandler\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mBaseHandler\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mHTTPError\u001b[0m: HTTP Error 403: FORBIDDEN"
     ]
    }
   ],
   "source": [
    "def describe(uri):\n",
    "    g = Graph()\n",
    "\n",
    "    for endpoint in sparql_endpoints:\n",
    "        sparql = SPARQLWrapper(endpoint)\n",
    "        sparql.setQuery(\"\"\"\n",
    "                DESCRIBE <\"\"\"+uri+\"\"\">\n",
    "            \"\"\")\n",
    "\n",
    "        sparql.setReturnFormat(N3)\n",
    "        results = sparql.query().convert()\n",
    "        g.parse(data=results, format=\"n3\")\n",
    "    \n",
    "    return g\n",
    "\n",
    "for uri in entities:\n",
    "    print(f'searching for {uri}')\n",
    "    augmented_kg = describe(uri)\n",
    "    print(augmented_kg.serialize(format='turtle').decode())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@prefix dash: <http://datashapes.org/dash#> .\n",
      "@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .\n",
      "@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .\n",
      "@prefix schema: <http://schema.org/> .\n",
      "@prefix sh: <http://www.w3.org/ns/shacl#> .\n",
      "@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .\n",
      "\n",
      "schema:DataCatalogShape\n",
      "    a sh:NodeShape ;\n",
      "    sh:targetClass schema:DataCatalog ;\n",
      "    sh:property [\n",
      "        sh:path schema:name ;\n",
      "        sh:datatype xsd:string ;\n",
      "        sh:name \"given name\" ;\n",
      "    ] ;\n",
      "    sh:property [\n",
      "        sh:path schema:description ;\n",
      "        sh:maxCount 1 ;\n",
      "    ] ;\n",
      "    sh:property [\n",
      "        sh:path schema:url ;\n",
      "    ] .\n",
      "\n",
      "11\n"
     ]
    }
   ],
   "source": [
    "# Exemple constraint Graph\n",
    "\n",
    "str_ttl_constraint = \"\"\"@prefix dash: <http://datashapes.org/dash#> .\n",
    "@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .\n",
    "@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .\n",
    "@prefix schema: <http://schema.org/> .\n",
    "@prefix sh: <http://www.w3.org/ns/shacl#> .\n",
    "@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .\n",
    "\n",
    "schema:DataCatalogShape\n",
    "    a sh:NodeShape ;\n",
    "    sh:targetClass schema:DataCatalog ;\n",
    "    sh:property [\n",
    "        sh:path schema:name ;\n",
    "        sh:datatype xsd:string ;\n",
    "        sh:name \"given name\" ;\n",
    "    ] ;\n",
    "    sh:property [\n",
    "        sh:path schema:description ;\n",
    "        sh:maxCount 1 ;\n",
    "    ] ;\n",
    "    sh:property [\n",
    "        sh:path schema:url ;\n",
    "    ] .\n",
    "\"\"\"\n",
    "\n",
    "shacl_graph = rdflib.ConjunctiveGraph()\n",
    "print(str_ttl_constraint)\n",
    "result = shacl_graph.parse(data=str_ttl_constraint, format='turtle')\n",
    "print(len(shacl_graph))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52\n"
     ]
    }
   ],
   "source": [
    "# Example data Graph\n",
    "data_graph_str = \"\"\"\n",
    "    {\n",
    "        \"@context\": \"http://schema.org\",\n",
    "        \"@type\": \"DataCatalog\",\n",
    "        \"identifier\":\"https://identifiers.org/MIR:00000364\",\n",
    "        \"name\": \"FAIRsharing.org\",\n",
    "        \"description\": \"A manually curated, informative and educational resource on data and metadata standards, inter-related to databases/data repositories and funder and journal publisher data policies from across disciplines. FAIRsharing is an ELIXIR-UK node resource and has an active role in the RDA and Force11 data initiatives.\",\n",
    "        \"url\": \"https://fairsharing.org/\",\n",
    "        \"dataset\": [\n",
    "            {\n",
    "                \"@type\" : \"Dataset\",\n",
    "                \"name\" : \"Metadata Standard\",\n",
    "                \"description\" : \"A manually curated registry of standards, split into three types - Terminology Artifacts (ontologies, e.g. Gene Ontology), Models and Formats (conceptual schema, formats, data models, e.g. FASTA), and Reporting Guidelines (e.g. the ARRIVE guidelines for in vivo animal testing). These are linked to the databases that implement them and the funder and journal publisher data policies that recommend or endorse their use.\",\n",
    "                \"identifier\": \"https://www.fairsharing.org/bsg-sXXXXXX\",\n",
    "                \"url\" : \"https://fairsharing.org/standards\",\n",
    "                \"keywords\" : \"Standards, Metadata, Formats, Ontologies, Terminology Artifacts, Reporting Guidelines\",\n",
    "                \"includedInDataCatalog\" : \"https://www.fairsharing.org\"\n",
    "            },\n",
    "            {\n",
    "                \"@type\" : \"Dataset\",\n",
    "                \"name\" : \"Database\",\n",
    "                \"description\": \"A manually curated registry of databases/data repositories, conforming to the BioDBcore standard (from the Life Sciences). These are linked to the standards that they use and the funder and journal publisher data policies that recommend or endorse their use.\",\n",
    "                \"identifier\": \"https://www.fairsharing.org/bsg-dXXXXXX\",\n",
    "                \"url\": \"https://fairsharing.org/databases\",\n",
    "                \"keywords\": \"Database, Data repository\",\n",
    "                \"includedInDataCatalog\" : \"https://www.fairsharing.org\"\n",
    "            },\n",
    "            {\n",
    "                \"@type\" : \"Dataset\",\n",
    "                \"name\" : \"Data Policy\",\n",
    "                \"description\": \"A manually curated registry of data policies from research funders, journal publishers, societies, and other organisations. These are linked to the databases and standards that they recommend for use\",\n",
    "                \"identifier\": \"https://www.fairsharing.org/bsg-pXXXXXX\",\n",
    "                \"url\": \"https://fairsharing.org/policies\",\n",
    "                \"keywords\": \"Data policy, journal, funder, society\"\n",
    "            }\n",
    "            ],\n",
    "        \"keywords\": \"registry, life science, natural science, social science, \",\n",
    "        \"provider\":\n",
    "            [\n",
    "            {\n",
    "            \"@type\":\"Person\",\n",
    "            \"name\": \"FAIRsharing.org support\",\n",
    "            \"email\": \"contact@fairsharing.org\"\n",
    "            },\n",
    "            {\n",
    "            \"@type\":\"Organization\",\n",
    "            \"name\": \"Oxford e-Research Centre, University of Oxford\",\n",
    "            \"url\": \"http://www.oerc.ox.ac.uk/\"\n",
    "            },\n",
    "            {\n",
    "            \"@type\":\"Organization\",\n",
    "            \"name\": \"ELIXIR-UK\",\n",
    "            \"url\": \"http://www.elixir-uk.org\"\n",
    "            },\n",
    "            {\n",
    "            \"alternateName\": [\"FAIRsharing.org Registry\", \"BioSharing.org\"]\n",
    "            }\n",
    "            ],\n",
    "        \"license\": {\n",
    "            \"@type\": \"CreativeWork\",\n",
    "            \"name\": \"Creative Commons CC BY-SA 4.0 Attribution\",\n",
    "            \"url\": \"https://creativecommons.org/licenses/by-sa/4.0/\"\n",
    "            },\n",
    "        \"publication\":[\n",
    "            {\n",
    "                \"@type\":\"PublicationEvent\",\n",
    "                \"name\": \"BioSharing: curated and crowd-sourced metadata standards, databases and data policies in the life sciences.\",\n",
    "                \"url\": \"https://academic.oup.com/database/article-lookup/doi/10.1093/database/baw075\"\n",
    "            }\n",
    "        ]\n",
    "\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "data_graph = rdflib.ConjunctiveGraph()\n",
    "result = data_graph.parse(data=data_graph_str, format='json-ld')\n",
    "print(len(data_graph))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "[a rdfg:Graph;rdflib:storage [a rdflib:Store;rdfs:label 'IOMemory']].\n",
      "Validation Report\n",
      "Conforms: True\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Exemple SHACL test\n",
    "\n",
    "r = pyshacl.validate(data_graph = data_graph, \n",
    "             data_graph_format='json-ld', \n",
    "             shacl_graph = shacl_graph, \n",
    "             #shacl_graph = my_shacl_constraint,\n",
    "             shacl_graph_format = 'turtle', \n",
    "             ont_graph = None, \n",
    "             inference = 'rdfs', \n",
    "             abort_on_error = False, \n",
    "             meta_shacl = False, \n",
    "             debug = True)\n",
    "\n",
    "conforms, results_graph, results_text = r\n",
    "\n",
    "print(conforms)\n",
    "print(results_graph)\n",
    "print(results_text)\n",
    "\n",
    "report_query = \"\"\"\n",
    "    SELECT ?node ?path WHERE {\n",
    "        ?v rdf:type sh:ValidationReport ;\n",
    "           sh:result ?r .\n",
    "        ?r sh:focusNode ?node ;\n",
    "           sh:sourceShape ?s . \n",
    "        ?s sh:path ?path . \n",
    "    }\n",
    "\"\"\"\n",
    "\n",
    "results = results_graph.query(report_query)\n",
    "\n",
    "for r in results :\n",
    "    print(r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "  {\n",
      "    \"@id\": \"https://w3id.org//FAIR_Tests/tests/gen2_unique_identifier#10.5281/zenodo.1147435/result-2020-04-20T15:33:51+00:00\",\n",
      "    \"http://purl.obolibrary.org/obo/date\": [\n",
      "      {\n",
      "        \"@value\": \"2020-04-20T15:33:51+00:00\",\n",
      "        \"@type\": \"http://www.w3.org/2001/XMLSchema#date\"\n",
      "      }\n",
      "    ],\n",
      "    \"http://schema.org/comment\": [\n",
      "      {\n",
      "        \"@value\": \"INFO: TEST VERSION 'Hvst-1.1.1:Tst-0.2.2'\\n\\nSUCCESS: Found an identifier of type 'doi'\",\n",
      "        \"@language\": \"en\"\n",
      "      }\n",
      "    ],\n",
      "    \"@type\": [\n",
      "      \"http://fairmetrics.org/resources/metric_evaluation_result\"\n",
      "    ],\n",
      "    \"http://semanticscience.org/resource/SIO_000300\": [\n",
      "      {\n",
      "        \"@value\": \"1\",\n",
      "        \"@type\": \"http://www.w3.org/2001/XMLSchema#int\"\n",
      "      }\n",
      "    ],\n",
      "    \"http://semanticscience.org/resource/SIO_000332\": [\n",
      "      {\n",
      "        \"@value\": \"10.5281/zenodo.1147435\",\n",
      "        \"@type\": \"http://www.w3.org/2001/XMLSchema#float\"\n",
      "      }\n",
      "    ],\n",
      "    \"http://schema.org/softwareVersion\": [\n",
      "      {\n",
      "        \"@value\": \"Hvst-1.1.1:Tst-0.2.2\",\n",
      "        \"@language\": \"en\"\n",
      "      }\n",
      "    ]\n",
      "  }\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "# FAIRMetrics API execution\n",
    "\n",
    "data = b'{\"subject\": \"10.5281/zenodo.1147435\"}'\n",
    "r = requests.request(method=\"POST\", url='https://w3id.org/FAIR_Tests/tests/gen2_unique_identifier', data=data)\n",
    "\n",
    "print(json.dumps(r.json(), indent=2))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
